### ISSUE TITLE
Need clarify with input / output sizes

### ISSUE THREAD

**Issue by venezuela01** [oxen-io/eth-sn-contracts#10]
1.

```
    /// @return r the sum of two points of G1
    function add(G1Point memory p1, G1Point memory p2) internal returns (G1Point memory r) {
        uint[4] memory input;
        input[0] = p1.X;
        input[1] = p1.Y;
        input[2] = p2.X;
        input[3] = p2.Y;
        bool success;
        assembly {
            success := call(sub(gas(), 2000), 6, 0, input, 0xc0, r, 0x60)
        }
        require(success, "Call to precompiled contract for add failed");
    }
```

`input` is unit[4], which is `32*4 = 128` bytes, which is 0x80 in hex
`r` is a G1Point which is `32*2 = 64` bytes, which is 0x40 in hex
I don't understand why we have `0xc0` and `0x60` in `call(sub(gas(), 2000), 6, 0, input, 0xc0, r, 0x60)`, is there anything I missed?

2.

```
    /// @return r the product of a point on G1 and a scalar, i.e.
    /// p == p.mul(1) and p.add(p) == p.mul(2) for all points p.
    function mul(G1Point memory p, uint s) internal returns (G1Point memory r) {
        uint[3] memory input;
        input[0] = p.X;
        input[1] = p.Y;
        input[2] = s;
        bool success;
        assembly {
            success := call(sub(gas(), 2000), 7, 0, input, 0x80, r, 0x60)
        }
        require(success, "Call to precompiled contract for mul failed");
    }
```

`input` is `unit[3]` which is `32*3 = 96` bytes or 0x60 in hex
`r` is G1Point which is `32*2 = 64` bytes or 0x40 in hex
I don't understand the line `call(sub(gas(), 2000), 7, 0, input, 0x80, r, 0x60)` either, can anyone explain it for me?

Source: https://github.com/oxen-io/eth-sn-contracts/blob/e3f6bdc4eef75bce44b44e076cae0e6163b046e4/contracts/libraries/BN256G1.sol

Thank you.

**Comment by darcys22** [comment]
This is simply passing the variables to the EVM precompiles. The ethereum clients have these curve operations built in

https://www.evm.codes/precompiled?fork=shanghai

**Comment by darcys22** [comment]
Oh youre asking about how the params to the call opcode work.

If i recall correctly the input and return objects are sent using RLP encoding. So its length is not simply concatenating the element sizes, its a “list” that has a prefix and a size. 

Plus there is some magic if the list is below a certain size.

So the input for add is 32 bytes prefix + 32 bytes length + 4x elements = 192 bytes

the input for mul is 32 bytes (prefix and length combined) + 3x elements = 128 bytes

the return for both is 32 bytes (prefix and length) and 2x elements = 96 bytes

**Comment by darcys22** [comment]
Deleted my other response cause I'm not sure anymore

ill need to investigate this further, this code has been copied and pasted and worked happily without any issues.

there is a chance that those size params arnt correct and just have had no effect on the functionality

**Comment by darcys22** [comment]
Yeah looks like that param is completely unused on the ethereum client.

Source was here
https://github.com/maticnetwork/pos-commit-poc

**Comment by KeeJef** [comment]
Full audit report will be released soon with findings, just finalising things now

**Comment by KeeJef** [comment]
My understanding is that they did not review any GitHub issues before determining their findings. We simply provided them with the commit hash, the repository link to the contracts folder, and a high-level outline of the contracts and interaction paths before the audit began. We did not provide any background on previously raised GitHub issues, and they did not bring up any community-identified issues with us.

We specifically directed Zellic to focus attention in their audit on our cryptographic implementation of BLS validation logic, which is where they spent most of their time and where we had the biggest blind spots, they also looked our other contracts, but those contracts implement much more basic functionality.
